satpy/readers/abi_base.py

```diff
@@ -0,0 +1,217 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Advance Baseline Imager reader base class for the Level 1b and l2+ reader
+"""
+
+import logging
+from datetime import datetime
+
+import numpy as np
+import xarray as xr
+
+from pyresample import geometry
+from satpy.readers.file_handlers import BaseFileHandler
+from satpy import CHUNK_SIZE
+
+logger = logging.getLogger(__name__)
+
+PLATFORM_NAMES = {
+    'G16': 'GOES-16',
+    'G17': 'GOES-17',
+}
+
+
+class NC_ABI_BASE(BaseFileHandler):
+
+    def __init__(self, filename, filename_info, filetype_info):
+        super(NC_ABI_BASE, self).__init__(filename, filename_info, filetype_info)
+        # xarray's default netcdf4 engine
+        try:
+            self.nc = xr.open_dataset(self.filename,
+                                      decode_cf=True,
+                                      mask_and_scale=False,
+                                      chunks={'x': CHUNK_SIZE, 'y': CHUNK_SIZE},)
+        except ValueError:
+            self.nc = xr.open_dataset(self.filename,
+                                      decode_cf=True,
+                                      mask_and_scale=False,
+                                      chunks={'lon': CHUNK_SIZE, 'lat': CHUNK_SIZE},)
+
+        self.nc = self.nc.rename({'t': 'time'})
+        platform_shortname = filename_info['platform_shortname']
+        self.platform_name = PLATFORM_NAMES.get(platform_shortname)
+        self.sensor = 'abi'
+
+        if 'goes_imager_projection' in self.nc.keys():
+            self.nlines = self.nc['y'].size
+            self.ncols = self.nc['x'].size
+        elif 'goes_lat_lon_projection' in self.nc.keys():
+            self.nlines = self.nc['lat'].size
+            self.ncols = self.nc['lon'].size
+            self.nc = self.nc.rename({'lon': 'x', 'lat': 'y'})
+
+        self.coords = {}
+
+    def __getitem__(self, item):
+        """Wrapper around `self.nc[item]`.
+
+        Some datasets use a 32-bit float scaling factor like the 'x' and 'y'
+        variables which causes inaccurate unscaled data values. This method
+        forces the scale factor to a 64-bit float first.
+
+        """
+        data = self.nc[item]
+        attrs = data.attrs
+        factor = data.attrs.get('scale_factor')
+        offset = data.attrs.get('add_offset')
+        fill = data.attrs.get('_FillValue')
+        if fill is not None:
+            data = data.where(data != fill)
+        if factor is not None:
+            # make sure the factor is a 64-bit float
+            # can't do this in place since data is most likely uint16
+            # and we are making it a 64-bit float
+            data = data * float(factor) + offset
+        data.attrs = attrs
+
+        # handle coordinates (and recursive fun)
+        new_coords = {}
+        # 'time' dimension causes issues in other processing
+        if 'time' in data.coords:
+            del data.coords['time']
```

Comment:
Some changes about this where introduced recently:

https://github.com/pytroll/satpy/pull/788/files#diff-57c2184527f32a105f3454c5f526c852